### ISSUE TITLE
Code Generation is too monolithic

### ISSUE THREAD

**Issue by ciaranmcnulty** [phpspec/phpspec#619]
Currently there are a few instances of code generation:

*Class generators*
* Generating specs on `describe`
* Generating classes when class does not exit

*Class modifiers*
* Generating methods that do not exist
* Generating return statements when `--fake` is used
* Generating named constructors

Future improvements are in the pipeline (Collaborator generation is one).

In most of these cases (aside from spec generation) the whole process consists of 2 components:

1. A Listener that detects code generation events and prompts the user whether they should proceed
2. A Generator that decides which template to use, invokes the template renderer, writes the changes to disk, then informs the user of success

It's clear that #2 at least is breaking SRP a few ways, the test suite and Scrutinizer are screaming about it.

**Comment by stof** [comment]
There is another PhpSpec library doing code generation: Prophecy to generate doubles. The main difference between them is that the Prophecy generator does not care about coding standards (given it is a temporary class) and so does not need to make the formatting configurable. Given these different requirements, I'm not sure there is any opportunity to share code, but this should be evaluated.

Regarding the needed refactoring, I think the first step (listening to events where generation may happen and then triggering the generator when needed) is fine.
However, the second step should indeed be split into several collaborators

**Comment by ciaranmcnulty** [comment]
The good news is that there are only `supports`/`getPriority`/`generate` methods in the interface. However these classes are non-final and contain protected methods, and are commonly overridden in Extensions, so care will need to be taken to retain BC.

**Comment by ciaranmcnulty** [comment]
@stof I think there is maybe an argument for a small standalone library for this sort of object definition manipulation that could be shared with projects like Prophecy

**Comment by stof** [comment]
yeah, this is what I meant when sayign we should evaluate whether a common codebase could be shared. I'm not sure it is possible though because of different needs

**Comment by gnugat** [comment]
Instead of putting the generation/edition logic in specialized classes, we could put it in templates (listeners would decide which one to render).
This would mean a better templating engine, maybe even a third party one, but this should be worth it.
That's the direction I've decided to go in [medio](https://github.com/gnugat/medio), after struggling with many "pretty printers", basically the same issue you're having (changes involving the templating are not yet available in my repo).

**Comment by Markcial** [comment]
+1, i am trying to write a phpspec extension, the plain strtr function had me disappointed a little :( i would like to render more complex snippets, like for example an array of params like "[id, name, email]" that would be converted to 
```
private $id;
private $name;
private $email;
```
would be great to have a templating engine plugged there or having a more complete templating capabilities.

**Comment by jakzal** [comment]
@Markcial @gnugat wrote an extension to integrate memio in phpspec: http://gnugat.github.io/2015/05/13/memio-spec-gen.html

**Comment by Markcial** [comment]
Oh thanks, i was going through this path actually, i will check this out :D

**Comment by gnugat** [comment]
@Markcial please note that the extension has a major downside: it doesn't support phpspec templates.
Memio templates can be customized (they are twig templates), but I still have to figure out a way to allow this in the extension.

**Comment by Markcial** [comment]
@gnugat I am using the fluent interface for the methods you provide in the api, there is still no need for templates  but the lack of use statements got me a little bit stuck, maybe i will do a PR to your repo if i get to understand how your library works.

**Comment by gnugat** [comment]
@Markcial That's because I wanted to release a Minimum Valuable Product for v0.1. Insertion of use statement is already on its way, should be released next week in v0.2 (I've already got a working prototype, I just need some time to finish the refactoring).

**Comment by Markcial** [comment]
I already found a method to set the uses with the fully wualified names, until the next release it works, but i will be tuned for the next release it  helps a lot creating the code in our behat and phpspec workflow :)
